src/server/frontend_wayland/wayland_connector.cpp

```diff
@@ -271,9 +274,34 @@ class WlCompositor : public wayland::Compositor::Global
     }
 };
 
+void WlCompositor::on_surface_created(wl_client* client, uint32_t id, std::function<void(WlSurface*)> const& callback)
+{
+    wl_resource* resource = wl_client_get_object(client, id);
+    auto const wl_surface = resource ? WlSurface::from(resource) : nullptr;
+    if (wl_surface)
+    {
+        callback(wl_surface);
+    }
+    else
+    {
+        surface_callbacks[std::make_pair(client, id)].push_back(callback);
+    }
+}
+
 void WlCompositor::Instance::create_surface(wl_resource* new_surface)
 {
-    new WlSurface{new_surface, compositor->executor, compositor->allocator};
+    auto const surface = new WlSurface{new_surface, compositor->executor, compositor->allocator};
+    auto const key = std::make_pair(wl_resource_get_client(new_surface), wl_resource_get_id(new_surface));
+    auto const callbacks = compositor->surface_callbacks.find(key);
```

Comment:
Shouldn't this be `callbacks = compositor->surface_callbacks.**equal_range**(key);` etc.?

Reply:
Actually, as surface_callbacks is a map that doesn't matter. (For some reason I was thinking it was a multimap, not a map with vector values)